### PR TITLE
[SPARK-11089] [SQL] Adds option for disabling multi-session in Thrift server

### PR DESCRIPTION
**Pull request by liancheng**
This PR adds a new option `spark.sql.hive.thriftServer.singleSession` for disabling multi-session support in the Thrift server.

Note that this option is added as a Spark configuration (retrieved from `SparkConf`) rather than Spark SQL configuration (retrieved from `SQLConf`). This is because all SQL configurations are session-ized. Since multi-session support is by default on, no JDBC connection can modify global configurations like the newly added one.

### CODE REVIEW

**Comment by liancheng** [comment]
Another approach for this problem is to have "global" options. For example, `SET -g key=value` sets `key` to `value` for all sessions. This adds a new concept and requires more work, but seems to be cleaner and in general more useful. What need to be done are:

1. Add a new syntax `SET -g`
1. Each `HiveContext` should remember its parent `HiveContext` so that we can retrieve the "root" one. 
1. When a global option is set, it's always set in the root `HiveContext`.

Another benefit of this approach is that we don't need to restart the Thrift server to switch between single-/multi-session modes.

@marmbrus @davies @yhuai Any suggestions?

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala [hunk 1]
```diff
@@ -635,6 +683,7 @@ abstract class HiveThriftServer2Test extends SparkFunSuite with BeforeAndAfterAl
        |  --driver-class-path $driverClassPath
        |  --driver-java-options -Dlog4j.debug
        |  --conf spark.ui.enabled=false
+       |  ${extraConf.mkString("\n")}
```
Comment (davies) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala]: Is it possible that we only introduce a command line option here? (Not sure how the JDBC server is used).

**Comment by davies** [comment]
@liancheng I like the this approach more, all global configurations should be placed in SparkConf (could be override in current session)

**Comment by davies** [comment]
@liancheng This option is only for backward compatibility purpose, could we document it somewhere? We should also mention this change (isolated sessions) in release note.

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala (hunk 1 above)
Comment (liancheng) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala]: Do you mean something like `--single-session`? The benefit of `SparkConf` option is that users can add it in spark-defaults.conf.

Command line options of `start-thriftserver.sh` are parsed by both `SparkSubmit` and `HiveServer2.ServerOptionsProcessor`. The latter isn't extensible. One option is to add a `--single-session` option in `SparkSubmit` and make a synonymous of `--conf spark.sql.hive.thriftServer.singleSession=true`. However, this option doesn't seem to be general enough to be added in `SparkSubmit`.

**Comment by liancheng** [comment]
Yeah, I'll add it to the programming guide.

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala (hunk 1 above)
Comment (davies) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Suites.scala]: I see, thanks!

**Comment by davies** [comment]
@liancheng Will you update the programming guide in this PR?

**Comment by liancheng** [comment]
Yeah, updating this one soon.

**Comment by marmbrus** [comment]
Merging to master and 1.6.

**Pull request closed without merging**